### PR TITLE
Support negative numbers as value

### PR DESCRIPTION
**Pull request by sundy-li**

### CODE REVIEW

**Comment by sundy-li** [comment]
Related https://github.com/datafuselabs/databend/pull/3197

File: src/ast/value.rs [hunk 1]
```diff
@@ -61,6 +63,37 @@ pub enum Value {
     Null,
 }
 
+impl Value {
+    pub fn to_negative(self) -> Self {
+        match self {
+            #[cfg(not(feature = "bigdecimal"))]
+            Value::Number(x, v) => Value::Number(format!("-{}", x), v),
```
Comment (sundy-li) [inline src/ast/value.rs:70]: @PsiACE  Do you why `format` is not allowed here but it can be allowed in other places.

Comment (PsiACE) [reply src/ast/value.rs:70]: `format!` allocates a `String`, and a `no_std` environment doesn't have an allocator.

https://stackoverflow.com/questions/50200268/how-can-i-use-the-format-macro-in-a-no-std-environment

Comment (sundy-li) [reply src/ast/value.rs:70]: but there are `format!` in other places of this repo.

Comment (PsiACE) [reply src/ast/value.rs:70]: i will fix it later

Comment (PsiACE) [reply src/ast/value.rs:70]: Done, use `alloc::format`

File: src/ast/value.rs [hunk 2]
```diff
@@ -11,13 +11,15 @@
...
-use alloc::string::String;
+use alloc::{format, string::String};
 use core::fmt;
 
 #[cfg(feature = "bigdecimal")]
 use bigdecimal::BigDecimal;
 #[cfg(feature = "serde")]
 use serde::{Deserialize, Serialize};
+#[cfg(feature = "bigdecimal")]
+use std::ops::Neg;
```
Comment (PsiACE) [inline src/ast/value.rs]: emmm, why not use `core::ops::Neg`

Comment (sundy-li) [reply src/ast/value.rs]: I did not notice this repo has no-std feature.

Comment (PsiACE) [reply src/ast/value.rs]: Done.

**Comment by dust1** [comment]
hello @sundy-li .don't i need to modify this?

**Comment by sundy-li** [comment]
> hello @sundy-li .don't i need to modify this?

You can modify the git version of `sqlparser`  crate in `databend` in your pr to verify datafuselabs/databend#3197

version:   b42ff01

**Comment by dust1** [comment]
ok

**Comment by dust1** [comment]
I edit two file: `common/exception/Cargo.tom`, `query/Cargo.tom`.

in my number type stateless both ok.
```
select toTypeName(a) from (select 42 as a);
select toTypeName(a) from (select -1 as a);
select toTypeName(a) from (select -128 as a);
select toTypeName(a) from (select -32768 as a);
select toTypeName(a) from (select -2147483648 as a);
select toTypeName(a) from (select -9223372036854775808 as a);
select toTypeName(a) from (select 255 as a);
select toTypeName(a) from (select 65535 as a);
select toTypeName(a) from (select 4294967295 as a);
select toTypeName(a) from (select 18446744073709551615 as a);
select toTypeName(a) from (select 1844674407370955161522 as a);
```
but `02_0014_function_maths` happen error, nohup.out:
```
[2021-12-05T12:57:00Z ERROR databend_query::servers::mysql::mysql_session] Unexpected error occurred during query execution: Code: 10, displayText = Unsupported column type:UInt16(while in query SELECT abs(-10086)).
    
       0: backtrace::backtrace::libunwind::trace
                 at /Users/kous/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/libunwind.rs:93:5
          backtrace::backtrace::trace_unsynchronized
                 at /Users/kous/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/mod.rs:66:5
          backtrace::backtrace::trace
                 at /Users/kous/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/mod.rs:53:14
       1: backtrace::capture::Backtrace::create
                 at /Users/kous/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/capture.rs:176:9
          backtrace::capture::Backtrace::new
                 at /Users/kous/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/capture.rs:140:22
       2: common_exception::exception::ErrorCode::BadDataValueType
                 at common/exception/src/exception.rs:113:77
       3: databend_query::servers::mysql::writers::query_result_writer::DFQueryResultWriter<W>::ok
                 at query/src/servers/mysql/writers/query_result_writer.rs:174:48
       4: databend_query::servers::mysql::writers::query_result_writer::DFQueryResultWriter<W>::write
                 at query/src/servers/mysql/writers/query_result_writer.rs:40:45
       5: <databend_query::servers::mysql::mysql_interactive_worker::InteractiveWorker<W> as msql_srv::MysqlShim<W>>::on_query
                 at query/src/servers/mysql/mysql_interactive_worker.rs:167:40
       6: msql_srv::MysqlIntermediary<B,R,W>::run
                 at /Users/kous/.cargo/git/checkouts/msql-srv-62b5153e0120bb3a/e4c8f3d/src/lib.rs:528:25
       7: msql_srv::MysqlIntermediary<B,R,W>::run_on
                 at /Users/kous/.cargo/git/checkouts/msql-srv-62b5153e0120bb3a/e4c8f3d/src/lib.rs:308:9
       8: msql_srv::MysqlIntermediary<B,std::net::tcp::TcpStream,std::net::tcp::TcpStream>::run_on_tcp
                 at /Users/kous/.cargo/git/checkouts/msql-srv-62b5153e0120bb3a/e4c8f3d/src/lib.rs:274:9
       9: databend_query::servers::mysql::mysql_session::MySQLConnection::session_executor
                 at query/src/servers/mysql/mysql_session.rs:44:29
      10: databend_query::servers::mysql::mysql_session::MySQLConnection::run_on_stream::{{closure}}
                 at query/src/servers/mysql/mysql_session.rs:35:13
      11: common_base::thread::Thread::spawn::{{closure}}
                 at common/base/src/thread.rs:32:17
      12: std::sys_common::backtrace::__rust_begin_short_backtrace
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/sys_common/backtrace.rs:123:18
      13: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/thread/mod.rs:484:17
      14: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/panic/unwind_safe.rs:271:9
      15: std::panicking::try::do_call
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panicking.rs:406:40
      16: <unknown>
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panicking.rs:434:6
      17: std::panicking::try
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panicking.rs:370:19
      18: std::panic::catch_unwind
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panic.rs:133:14
      19: std::thread::Builder::spawn_unchecked::{{closure}}
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/thread/mod.rs:483:30
      20: core::ops::function::FnOnce::call_once{{vtable.shim}}
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/ops/function.rs:227:5
      21: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/alloc/src/boxed.rs:1694:9
          <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/alloc/src/boxed.rs:1694:9
          std::sys::unix::thread::Thread::new::thread_start
                 at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/sys/unix/thread.rs:108:17
      22: __pthread_body
      23: __pthread_start
```
At the same time, the result of `06_0000_show_queries` shows the output of other strings
```
06_0000_show_queries:                                                   [ FAIL ] - result differs with:
--- /Users/kous/myProjects/rustproject/databend/tests/suites/0_stateless/06_0000_show_queries.result    2021-10-27 21:48:32.000000000 +0800
+++ /Users/kous/myProjects/rustproject/databend/tests/suites/0_stateless/06_0000_show_queries.stdout    2021-12-05 20:57:10.000000000 +0800
@@ -1,3 +1,4 @@
+math_sample_numbers
 t1
 t2
 t3
@@ -8,6 +9,7 @@
 t1
 t2
 t3
+math_sample_numbers
 t1
 t2
 t3
```

**Comment by sundy-li** [comment]
You can debug this by printing `data_type` in `https://github.com/datafuselabs/databend/blob/1600ee1e678c12cff6fe41ad5861bac491a29be5/query/src/servers/mysql/writers/query_result_writer.rs:176`  

Then `select abs(-10086)`

**Comment by dust1** [comment]
error data_value : 10086, data_type: UInt16

**Pull request merged**